lib/Dialect/Arc/Transforms/SimplifyVariadicOps.cpp

```diff
@@ -0,0 +1,127 @@
+//===- SimplifyVariadicOps.cpp --------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arc-simplify-variadic-ops"
+
+using namespace mlir;
+using namespace circt;
+using namespace arc;
+using namespace hw;
+
+namespace {
+struct SimplifyVariadicOpsPass
+    : public SimplifyVariadicOpsBase<SimplifyVariadicOpsPass> {
+  SimplifyVariadicOpsPass() = default;
+  SimplifyVariadicOpsPass(const SimplifyVariadicOpsPass &pass)
+      : SimplifyVariadicOpsPass() {}
+
+  void runOnOperation() override;
+  void simplifyOp(Operation *op);
+
+  Statistic numOpsSkippedMultipleBlocks{
+      this, "skipped-multiple-blocks",
+      "Ops skipped due to operands in different blocks"};
+  Statistic numOpsSimplified{this, "simplified",
+                             "Ops simplified into binary ops"};
+  Statistic numOpsCreated{this, "created",
+                          "Ops created as part of simplification"};
+  Statistic numOpsReordered{this, "reordered",
+                            "Ops where simplification reordered operands"};
+};
+} // namespace
+
+void SimplifyVariadicOpsPass::runOnOperation() {
+  SmallVector<Operation *> opsToProcess;
+  getOperation().walk([&](Operation *op) {
+    if (op->hasTrait<OpTrait::IsCommutative>() && op->getNumRegions() == 0 &&
+        op->getNumSuccessors() == 0 && op->getNumResults() == 1 &&
+        op->getNumOperands() > 2)
```

Comment:
nit: Can we check op has no side-effect?

Reply:
I think to be precise we would also have to check for associativity. E.g. imagine an operation `set_create` similar to `array_create`, but where the order of the operands does not matter (therefore you might want to add the commutativity trait), but you cannot pull the variadic list apart (not sure if an op with the same properties already exists or not).

Reply:
Yeah, I don't think there is a trait for associativity property. ExportVerilog also lowers variadic operations into binary ones (https://github.com/llvm/circt/blob/main/lib/Conversion/ExportVerilog/PrepareForEmission.cpp#L798-L820) so maybe we can share some helper function or create a trait for that

Reply:
Yeah having some op trait for this would be great! Probably something that's worth adding to MLIR upstream?